internal/catalog/catalog.go

```diff
@@ -166,16 +167,96 @@ func untar(dst, version string, tektonResources map[string]contract.TektonResour
 			// to wait until all operations have completed.
 			f.Close()
 
-			if filename != "README.md" {
-				if tektonResource.Checksum != sum {
-					fmt.Fprintf(os.Stderr, "%s checksum is different than the specified checksum in the catalog file: %s", sum, tektonResource.Checksum)
-					// FIXME: maybe handle *all* file before erroring out ?
-					return fmt.Errorf("invalid checksum for %s: %s != %s", filename, sum, tektonResource.Checksum)
-				}
-				fmt.Fprintf(os.Stderr, "✅ %s\n", tektonResource.Filename)
+					if filename != "README.md" {
+							if tektonResource.Checksum != sum {
+									fmt.Fprintf(os.Stderr, "%s checksum is different than the specified checksum in the catalog file: %s", sum, tektonResource.Checksum)
+									// FIXME: maybe handle *all* file before erroring out ?
+									return fmt.Errorf("invalid checksum for %s: %s != %s", filename, sum, tektonResource.Checksum)
+							}
+							fmt.Fprintf(os.Stderr, "✅ %s\n", tektonResource.Filename)
+					}
+
+					// Add "source" annotation to task YAML file
+					if strings.HasSuffix(target, ".yaml") {
+							if err := addSourceAnnotationToTask(target, resourcesURI); err != nil {
+									return err
+							}
+					}
 			}
+	}
+	
+}
+
+func addSourceAnnotationToTask(file, resourcesURI string) error {
+	// Add the new annotation
+	repoURL := extractRepositoryURL(resourcesURI)
+	// Open the Task YAML file
+	f, err := os.OpenFile(file, os.O_RDWR, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Create a scanner to read the file line by line
+	scanner := bufio.NewScanner(f)
+	var updatedContent []string
+
+	// Regular expression pattern to match the annotations in Task metadata
+	pattern := regexp.MustCompile(`^\s+annotations:\s*$`)
+
+	// Flag to indicate if the "tekton.dev/source" annotation is already present
+	var sourceAnnotationExists bool
+
+	// Read the file line by line
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// Check if the line matches the annotations pattern
+		if pattern.MatchString(line) {
+			// If annotations are found, set the flag to true
+			sourceAnnotationExists = true
 		}
+
+		// Append the line to updatedContent slice
+		updatedContent = append(updatedContent, line)
+
+		// Check if the "tekton.dev/source" annotation needs to be added
+		if sourceAnnotationExists && !pattern.MatchString(line) {
+			// Add the source annotation in the desired format
+			updatedContent = append(updatedContent, fmt.Sprintf("    tekton.dev/source: \"%s\"", repoURL))
+			// Reset the flag after adding the annotation
+			sourceAnnotationExists = false
+		}
+	}
+
+	// Check for scanner errors
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+
+	// Clear the file content and write the updated content
+	if err := f.Truncate(0); err != nil {
+		return err
+	}
+	if _, err := f.Seek(0, 0); err != nil {
+		return err
+	}
+	writer := bufio.NewWriter(f)
+	for _, line := range updatedContent {
+		fmt.Fprintln(writer, line)
+	}
```

Comment:
Do we require this ?

Reply:
Essentially, these lines of code ensure that the modified content stored in the updatedContent slice is written back to the Task YAML file effectively updating its contents with the added tekton.dev/source annotation. Without them the build test will fail as the generated file won't have any output to show

Reply:
Yeah we require these lines to write back the updated content from the slice back to the Task. Without them we would face the similar issue as before i.e. generated state doesn't match the expected state hence test will fail.